core/util/src/main/java/org/alice/math/immutable/Matrix4x4.java

```diff
@@ -0,0 +1,254 @@
+package org.alice.math.immutable;
+
+import edu.cmu.cs.dennisc.math.EpsilonUtilities;
+
+public interface Matrix4x4 {
+  AffineMatrix4x4 IDENTITY = new AffineMatrix4x4(Matrix3x3.IDENTITY, Vector3.ZERO);
+  AffineMatrix4x4 NaN = new AffineMatrix4x4(Matrix3x3.NaN, Vector3.NaN);
+
+  boolean isAffine();
+  boolean isNaN();
+  boolean isIdentity();
+
+  boolean isWithinEpsilonOf(Matrix4x4 other, double epsilon);
+
+  default boolean isWithinReasonableEpsilonOf(Matrix4x4 other) {
+    return isWithinEpsilonOf(other, EpsilonUtilities.REASONABLE_EPSILON);
+  }
+
+  static Matrix4x4 create(double e11, double e12, double e13, double e14, double e21, double e22, double e23, double e24, double e31, double e32, double e33, double e34, double e41, double e42, double e43, double e44) {
+    // Comes in as row major values
+    if (e41 == 0 && e42 == 0 && e43 == 0 && e44 == 1.0) {
+      // Affine matrix always have the same 4th row
+      // It stores the rest as a 3x3 orientation plus a translation vector
+      return new AffineMatrix4x4(
+          Matrix3x3.create(e11, e12, e13, e21, e22, e23, e31, e32, e33),
+          new Vector3(e14, e24, e34));
+    }
+    // FullMatrix stores a column in each vector
+    return new FullMatrix4x4(
+        new Vector4(e11, e21, e31, e41),
+        new Vector4(e12, e22, e32, e42),
+        new Vector4(e13, e23, e33, e43),
+        new Vector4(e14, e24, e34, e44));
+  }
+
+  static Matrix4x4 fromTranslation(Vector3 p) {
+    return new AffineMatrix4x4(Matrix3x3.IDENTITY, p);
+  }
+
+  static Matrix4x4 fromScale(double x, double y, double z) {
+    return new AffineMatrix4x4(
+        new Matrix3x3(
+            new Vector3(x, 0, 0),
+            new Vector3(0, y, 0),
+            new Vector3(0, 0, z)),
+        Vector3.ZERO);
+  }
+
```

Comment:
I'm not sure if now is the best or worst time to ask this- but i'd love to change these transforms to work on tuples instead of points and vectors 😇

Reply:
really, i think what i was dreaming of was a <T extends Tuple4>

Reply:
I am with you on intent. I ran into trouble when I tried that way was creating a `new T(x, y, z)` and went back to this.
I also considered dispatching though the Tuple3 classes (`Tuple3.transformBy(matrix)`), so they would create the instances of their own type, but that led to a generic interim value or functions for each value `transformX()`, `transformY()`, …
I remain unsatisfied.